commands/update_bootcfg/update_bootcfg.8
.\" Written by Antoine Leca
.Dd March 28, 2013
.Dt UPDATE_BOOTCFG 8
.Os MINIX
.Sh NAME
.Nm update_bootcfg
.Nd Create or udpate the configuration for the boot monitor
.Sh SYNOPSIS
.Nm
.Op Fl lx
.Op Fl r Ar rootdev
.Op Fl o Ar target
.Sh DESCRIPTION
The
.Nm
command recreates the configuration file used by
MINIX boot monitor. It is created from two templates,
.Pa /etc/boot.cfg.default
and
.Pa /etc/boot.cfg.local ;
the former is updated when the system changes,
while the latter holds the local modifications.
The templates obey the syntax for
.Xr boot.cfg 5 .
.Pp
The
.Fl o
option specifies the name of the new configuration file;
the default value is to update
.Pa /boot.cfg .
.Pp
The
.Pa /etc/boot.cfg.default
template expects the system images to reside in the
.Pa /boot/minix
directory, and that
.Pa /boot/minix_latest
points to last image, selected by the
.Cm default=
command. The
.Fl x
option will cause the former lastest image to be dropped,
and the
.Fl l
option will reset
.Pa /boot/minix_latest
to point to most recent image (measured with the time stamp of the
.Va kernel
file.) Besides the directory to which points
.Pa /boot/minix_latest ,
all the directory entries within 
.Pa /boot/minix
are expected to be alternative system releases, and are given additional
.Em menu=
entries, with the basic parameters
.Va $rootdevname
and
.Va $args .
Template
.Cm menu=
entries for inexistant directories, or directories without
.Va kernel
file, are skipped; the
.Cm default=
option is adjusted to point to the same relative menu entry
in the original template. This allows a
.Cm default=
pointing to a local
.Cm menu=
in the 
.Pa /etc/boot.cfg.local
template to take precedence in the resulting
.Pa /boot.cfg ;
the
.Cm clear=
and
.Cm timeout=
values in the
.Pa /etc/boot.cfg.local
template also take preference.
.Pp
Within the templates, some variables can be used and will be
substituted by
.Nm
at run time. They are preceded by the dollar sign.
.Sy $rootdevname
is replaced with the name of the device holding the
root file system. What is considered the root file system is
determined from its device name, as indicated either by the value
of the
.Sy $ROOT
setting in the
.Pa /etc/boot.cfg.args
file if set, else from the currently mounted device;
this can be overriden with the
.Fl r
option, for example if / is mounted as ramdisk.
.Sy $args ,
which is intended to be at the end of the menu lines, is replaced
by the system environement values, as determined at install time
by the
.Xr setup 8
utility and stored in
.Pa /etc/boot.cfg.args
file.
.Pp
The
.Nm
command updates the root file system, which it assumes to be the file
system the
.Xr boot 8
monitor starts from; also it reads the configuration files from the
.Pa /etc
directory; both references to the root can be changed using
.Bd -unfilled -option indent -compact
.Cm chroot \fIdestdir\fB update_bootcfg \fIoptions...\fB
.Ed
like to operate on a different file system; more complex cases can
be handled setting different values for the
.Sy $boot
and
.Sy $mboot
environment variables; see the beginning of the script for details.
.Sh FILES
.Bl -tag -width /etc/boot.cfg.default -compact
.It Pa /boot.cfg
Location of the configuration file; previous content is lost
.It Pa /boot/minix/*
Directories with the various releases
.It Pa /boot/minix_default
Directory with the base release of MINIX, installed at
.Xr setup 8
time.
.It Pa /boot/minix_latest
Symbolic link to the directory with the latest release, usually
taken by default
.It Pa /etc/boot.cfg.args
Locally-defined system environment variables, initialised at
.Xr setup 8
time.
.It Pa /etc/boot.cfg.default
Default template
.It Pa /etc/boot.cfg.local
Locally-defined additional boot commands
.El
.Sh BUGS
.Pp
The presence of
.Pa /boot/minix_latest
as a
.Cm menu=
entry is not checked; it is assumed than one or several entries refer to it.
.Pp
While the script is intended to be portable among POSIX implementations,
it still uses
.Xr stat 1
and
.Xr readlink 1
which are not defined in the Standard.
.Pp
The
.Nm
does not provide a
.Fl N
option (similar to
.Xr install 1
or
.Xr mtree 8 )
to override the place where
.Pa /etc
configuration files are searched.
.Pp
The naming conventions, particularly the use of
.Em "default" ,
are not consistent.
.Sh SEE ALSO
.Xr boot 8 ,
.Xr boot.cfg 5 ,
.Xr setup 8 .


commands/update_bootcfg/update_bootcfg.sh
#!/bin/sh
set -e

# Default values can be overriden from environment
# Files used by this utility:
: ${ARGSCFG:=/etc/boot.cfg.args}
: ${DEFAULTCFG:=/etc/boot.cfg.default}
: ${LOCALCFG:=/etc/boot.cfg.local}
# Where the boot monitor will look for its files?
# can be "" for a separate boot partition, or "hd1i:/boot" for 2nd disk
: ${boot=/boot}
# Where are currently the boot monitor files?
# examples can be "/mnt/boot" or "$DESTDIR/boot"
: ${mboot:=/boot}
: ${BOOTCFG:=/boot.cfg}
# Filepaths as seen from the boot monitor $boot, or from $mboot here:
: ${DIRSBASE=/minix} # Without final /, please!
: ${minix_default:=/minix_default}
: ${minix_latest:=/minix_latest}

# This script aims at being portable among any POSIX shell, in order to be
# run in cross-building cases (with chroot or jails.)
# XXX Currently makes use of readlink(1) and stat(1)

filter_entries()
{
	# This routine performs three tasks:
	# - substitute variables in the configuration lines;
	# - remove multiboot entries that refer to nonexistent kernels;
	# - adjust the default option for line removal and different files.
	# The last part is handled by the awk part of the routine.

	while read line
	do
		# Substitute variables like $boot, $rootdevname, and $args
		line=$(eval echo \"$line\")

		if ! echo "$line" | grep -sq '^menu=.*multiboot'
		then
			echo "$line"
			continue
		fi

		# Check if the referenced kernel is present
		kernel=`echo "$line" | sed -n 's!.*multiboot[[:space:]]*'$boot'\(/[^[:space:];]*\).*!\1!p'`
		if [ ! -r "$mboot$kernel" ]
		then
			echo "Warning: template contains entry for \"$boot$kernel\" which is missing! Entry skipped." 1>&2
			echo "menu=SKIP"
		else
			echo "$line"
		fi
	done | awk '
		BEGIN {
			count=1
			base=0
			default=0
		}
		/^menu=SKIP$/ {
			# A menu entry discarded by the kernel check above
			skip[count++]=1
			next
		}
		/^menu=/ {
			# A valid menu entry
			skip[count++]=0
			print
			next
		}
		/^BASE=/ {
			# The menu count base as passed in by count_entries()
			sub(/^.*=/,"",$1)
			base=$1+0
			next
		}
		/^default=/ {
			# The default option
			# Correct for the menu count base and store for later
			sub(/^.*=/,"",$1)
			default=$1+base
			next
		}
		{
			# Any other line
			print
		}
		END {
			# If a default was given, correct for removed lines
			# (do not bother to warn if the default itself is gone)
			if (default) {
				for (i = default; i > 0; i--)
					if (skip[i]) default--;
				print "default=" default
			}
		}
	'
}

count_entries()
{
	printf "BASE="; grep -cs '^menu=' "$1"
}

usage()
{
	cat >&2 <<'EOF'
Usage:
  update_bootcfg [-xl] [-r rootdev] [-o dest]

  Recreates the configuration file used by MINIX boot monitor.

  Options:
     -x Remove "minix_latest/*", for example if broken; then fix the link
     -l Restore "minix_latest" to point to lastest kernel
     -r Root device name; default is current device for /
     -o New configuration file to create; default is $BOOTCFG
EOF
	exit 1
}

while getopts "lo:r:x" c
do
	case "$c" in
	l)	do_restore_latest=1 ;;
	o)	keepexist=1; BOOTCFGTMP="$OPTARG" ;;
	r)	ROOT="$OPTARG" ;;
	x)	do_remove_latest=1 ;;
	*)	usage ;;
	esac
done
shift `expr $OPTIND - 1`

# Let see if ROOT was initialized by setup:
if test -z "$ROOT" -a -r "$ARGSCFG"
then
	if grep -q "^ROOT=" $ARGSCFG
	then
		: ${ROOT:-$(sed -n "s/^ROOT=//p" $ARGSCFG)}
	fi
fi
# Last chance: search which device holds the current / file system:
if test -z "$ROOT"
then
	ROOT=$(awk '$3=="/" { print $1; }' /etc/mtab)
	if test "x$ROOT" = "x/dev/ram"
	then # The root device is the ramdisk!
		# Avoid $(sysenv ramimagename) as being MINIX-specific
		echo Warning: \$ROOT is a ramdisk; consider -r option
	fi
fi

if [ ! -b "$ROOT" ]
then
	echo root device $ROOT not found
	exit 1
fi

rootdevname=`echo $ROOT | sed 's/\/dev\///'`

# Construct a list of additional arguments for boot options to use, based
# on a user-editable file /etc/boot.cfg.args initialized during setup.
# Note that rootdevname is not be passed on this way, as it is handled
# manually by this script to allow menu lines like
#   menu=Root FS in RAM: [...] ;multiboot $boot$minix_latest/kernel \
#           rootdevname=ram ramimagename=$rootdevname $args
# Also, exclude the ROOT variable which is handled above.
if test -r "$ARGSCFG"
then
	args=$( egrep -v "^ROOT=|^#" $ARGSCFG | xargs echo )
fi

# Remove the directory pointed to as /boot/minix_latest
# Useful when one realizes the last release does not work...
if test -n "$do_remove_latest"
then
	rm -rf $(readlink -f $mboot$minix_latest)
	[ ! -h $mboot$minix_latest ] || rm $mboot$minix_latest

	do_restore_latest=1
fi

# Restore /boot/minix_latest pointing at the directory with the lastest kernel
if test -n "$do_restore_latest"
then
	[ ! -h $mboot$minix_latest ] || rm $mboot$minix_latest

	target=$(ls -t $mboot$DIRSBASE/*/kernel 2>/dev/null \
		| sed -ne '1s|.*/\([^/]*\)/kernel$|\1|p')
	if test -z "$target" -o ! -d $mboot$DIRSBASE/"$target"
	then
		echo Warning! No MINIX kernels found at $mboot$DIRSBASE, cannot set $minix_latest>&2
		return
	fi
	# XXX Consider a warning if test "$target" = ".temp"

	# XXX Blindly assume $DIRSBASE and $minix_latest are sharing the same base
	ln -s $(basename $DIRSBASE)/"$target" $mboot$minix_latest
fi

# All is in place; let's proceed
[ -n "$keepexist" ] || BOOTCFGTMP=$BOOTCFG.temp

echo "# Generated file. Edit $LOCALCFG and run $0 !" > $BOOTCFGTMP

if test -r "$DEFAULTCFG"
then
	echo \# Template is "$DEFAULTCFG" >> $BOOTCFGTMP
	filter_entries < $DEFAULTCFG >> $BOOTCFGTMP
fi

if [ -d $mboot$minix_latest -o -h $mboot$minix_latest ]
then
	latest=$(basename $(stat -f "%Y" $mboot$minix_latest))
fi

[ -d $mboot$DIRSBASE ] && for i in `ls -tr $mboot$DIRSBASE/`
do
	build_name="`basename $i`"
	if [ "$build_name" != "$latest" ]
	then
	# Avoid inserting space before commands, boot monitor dislikes it
		echo "menu=Start MINIX 3 ($build_name):load_mods" \
			"$boot$DIRSBASE/$i/mod*;multiboot $boot$DIRSBASE/$i/kernel" \
				"rootdevname=$rootdevname $args" >> $BOOTCFGTMP
	fi
done

if test -r "$LOCALCFG"
then
	echo \# Local options from "$LOCALCFG" >> $BOOTCFGTMP

	# If the local config supplies a "default" option, we assume that this
	# refers to one of the options in the local config itself. Therefore,
	# we increase this default by the number of options already present in
	# the output so far. To this end, count_entries() inserts a special
	# token that is recognized and filtered out by filter_entries().
	(count_entries $BOOTCFGTMP; cat $LOCALCFG) | filter_entries >> $BOOTCFGTMP
fi

[ -n "$keepexist" ] || mv $BOOTCFGTMP $BOOTCFG

sync || true

